tests/std/tests/P2162R2_std_visit_for_derived_classes_from_variant/test.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <memory>
+#include <utility>
+#include <variant>
+
+using namespace std;
+
+struct Disconnected {
+    int val;
+};
+
+struct Connecting {
+    char val;
+};
+
+struct Connected {
+    double val;
+};
+
+struct State : variant<Disconnected, Connecting, Connected> {
+    using variant::variant;
+};
+
+void example1_from_p2162r2() {
+    State v1       = Disconnected{45};
+    const State v2 = Connecting{'d'};
+    visit([](auto x) { assert(x.val == 45); }, v1);
+    visit([](auto x) { assert(x.val == 'd'); }, v2);
+    visit([](auto x) { assert(x.val == 5.5); }, State{Connected{5.5}});
+    visit([](auto x) { assert(x.val == 45); }, move(v1));
+    visit([](auto x) { assert(x.val == 'd'); }, move(v2));
+}
+
+struct Expr;
+
+struct Neg {
+    shared_ptr<Expr> expr;
+};
+
+struct Add {
+    shared_ptr<Expr> lhs, rhs;
+};
+
+struct Mul {
+    shared_ptr<Expr> lhs, rhs;
+};

```

Comment:
We conventionally avoid defining multiple variables (or data members) on a single line. I see that this was taken directly from the paper's example, but we can change this without affecting the example's behavior (the code will still strongly resemble the original paper).